From fa6fb17c104589ef3345c2462787a4cc00331bf9 Mon Sep 17 00:00:00 2001
From: zhangshuangshuang <zhangshuangshuang@ict.ac.cn>
Date: Sun, 5 Jun 2016 16:31:29 +0800
Subject: [PATCH 086/130] Loongson-3: Add /proc/boardinfo support for
 Loongson3A/B(2000)

Some applications such as X with GC1000 GPU use /proc/boardinfo to determine
actions should be taken, this pacth make those applications happy.

Change-Id: Id3f50c715cbc8dca365e283138175c9ea33cee20
Signed-off-by: Zhang Shuangshuang <zhangshuangshuang@ict.ac.cn>
---
 arch/mips/loongson64/common/Makefile    |  3 +-
 arch/mips/loongson64/common/boardinfo.c | 77 +++++++++++++++++++++++++++++++++
 arch/mips/loongson64/common/env.c       | 32 +++++++++++++-
 3 files changed, 110 insertions(+), 2 deletions(-)
 create mode 100644 arch/mips/loongson64/common/boardinfo.c

diff --git a/arch/mips/loongson64/common/Makefile b/arch/mips/loongson64/common/Makefile
index 7838e345..27908018 100644
--- a/arch/mips/loongson64/common/Makefile
+++ b/arch/mips/loongson64/common/Makefile
@@ -3,7 +3,8 @@
 #
 
 obj-y += setup.o init.o cmdline.o env.o time.o reset.o irq.o \
-    bonito-irq.o mem.o machtype.o platform.o serial.o efi.o
+    bonito-irq.o mem.o machtype.o platform.o serial.o efi.o  \
+    boardinfo.o
 obj-$(CONFIG_PCI) += pci.o
 
 #
diff --git a/arch/mips/loongson64/common/boardinfo.c b/arch/mips/loongson64/common/boardinfo.c
new file mode 100644
index 00000000..562829a8
--- /dev/null
+++ b/arch/mips/loongson64/common/boardinfo.c
@@ -0,0 +1,77 @@
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/proc_fs.h>
+#include <linux/seq_file.h>
+#include <linux/module.h>
+
+#include <boot_param.h>
+
+char *bios_vendor;
+char *bios_release_date;
+char *board_manufacturer;
+char _bios_info[64];
+char _board_info[64];
+extern struct interface_info *einter;
+extern struct board_devices *eboard;
+
+static int show_boardinfo(struct seq_file *m, void *v)
+{
+	seq_printf(m, "BIOS Information\n");
+	seq_printf(m, "Vendor\t\t\t: %s\n", bios_vendor);
+	seq_printf(m, "Version\t\t\t: %s\n", einter->description);
+	seq_printf(m, "BIOS ROMSIZE\t\t: %d\n", einter->size);
+	seq_printf(m, "Release date\t\t: %s\n", bios_release_date);
+	seq_printf(m, "\n");
+
+	seq_printf(m, "Base Board Information\t\t\n");
+	seq_printf(m, "Manufacturer\t\t: %s\n", board_manufacturer);
+	seq_printf(m, "Board name\t\t: %s\n", eboard->name);
+	seq_printf(m, "Family\t\t\t: LOONGSON3\n");
+	seq_printf(m, "\n");
+
+	return 0;
+}
+
+static void *bd_start(struct seq_file *m, loff_t *pos)
+{
+	unsigned long i = *pos;
+
+	return i ? NULL : (void *)0x1;
+}
+
+static void *bd_next(struct seq_file *m, void *v, loff_t *pos)
+{
+	++*pos;
+	return bd_start(m, pos);
+}
+
+static void bd_stop(struct seq_file *m, void *v)
+{
+}
+
+const struct seq_operations boardinfo_op = {
+	.start	= bd_start,
+	.next	= bd_next,
+	.stop	= bd_stop,
+	.show	= show_boardinfo,
+};
+
+static int boardinfo_open(struct inode *inode, struct file *file)
+{
+	return seq_open(file, &boardinfo_op);
+}
+
+static const struct file_operations proc_boardinfo_operations = {
+	.open		= boardinfo_open,
+	.read		= seq_read,
+	.llseek		= seq_lseek,
+	.release	= seq_release,
+};
+
+static int __init proc_boardinfo_init(void)
+{
+	proc_create("boardinfo", 0, NULL, &proc_boardinfo_operations);
+	return 0;
+}
+
+module_init(proc_boardinfo_init);
diff --git a/arch/mips/loongson64/common/env.c b/arch/mips/loongson64/common/env.c
index c28e8a1a..1ed58801 100644
--- a/arch/mips/loongson64/common/env.c
+++ b/arch/mips/loongson64/common/env.c
@@ -44,6 +44,16 @@ struct platform_controller_hub *loongson_pch;
 extern struct platform_controller_hub ls2h_pch;
 extern struct platform_controller_hub rs780_pch;
 
+struct board_devices *eboard;
+struct interface_info *einter;
+struct loongson_special_attribute *especial;
+
+extern char *bios_vendor;
+extern char *bios_release_date;
+extern char *board_manufacturer;
+extern char _bios_info[];
+extern char _board_info[];
+
 #define parse_even_earlier(res, option, p)				\
 do {									\
 	unsigned int tmp __maybe_unused;				\
@@ -66,6 +76,8 @@ void __init prom_init_env(void)
 {
 	/* pmon passes arguments in 32bit pointers */
 	unsigned int processor_id;
+        char *bios_info;
+        char *board_info;
 
 #ifndef CONFIG_LEFI_FIRMWARE_INTERFACE
 	int *_prom_envp;
@@ -92,7 +104,6 @@ void __init prom_init_env(void)
 	struct boot_params *boot_p;
 	struct loongson_params *loongson_p;
 	struct system_loongson *esys;
-	struct board_devices *eboard;
 	struct efi_cpuinfo_loongson *ecpu;
 	struct irq_source_routing_table *eirq_source;
 
@@ -106,10 +117,14 @@ void __init prom_init_env(void)
 		((u64)loongson_p + loongson_p->cpu_offset);
 	eboard	= (struct board_devices *)
 		((u64)loongson_p + loongson_p->boarddev_table_offset);
+        einter = (struct interface_info *)
+                ((u64)loongson_p + loongson_p->interface_offset);
 	eirq_source = (struct irq_source_routing_table *)
 		((u64)loongson_p + loongson_p->irq_offset);
 	loongson_memmap = (struct efi_memory_map_loongson *)
 		((u64)loongson_p + loongson_p->memory_offset);
+        especial = (struct loongson_special_attribute *)
+                ((u64)loongson_p + loongson_p->special_offset);
 
 	cpu_clock_freq = ecpu->cpu_clock_freq;
 	loongson_sysconf.cputype = ecpu->cputype;
@@ -212,6 +227,21 @@ void __init prom_init_env(void)
 			loongson_sysconf.pci_io_base = 0x1ff00000;
         }
 
+        /* parse bios info */
+        strcpy(_bios_info, einter->description);
+        bios_info = _bios_info;
+        bios_vendor = strsep(&bios_info, "-");
+        strsep(&bios_info, "-");
+        strsep(&bios_info, "-");
+        bios_release_date = strsep(&bios_info, "-");
+        if (!bios_release_date)
+                bios_release_date = especial->special_name;
+
+        /* parse board info */
+        strcpy(_board_info, eboard->name);
+        board_info = _board_info;
+        board_manufacturer = strsep(&board_info, "-");
+
 	loongson_sysconf.restart_addr = boot_p->reset_system.ResetWarm;
 	loongson_sysconf.poweroff_addr = boot_p->reset_system.Shutdown;
 	loongson_sysconf.suspend_addr = boot_p->reset_system.DoSuspend;
-- 
2.11.0

